torch/distributed/rpc.py

```diff
@@ -158,6 +159,105 @@ def remote(to, func, args=None, kwargs=None):
+        func (callable): any callable function. builtin functions (like
+                         ``torch.add``) can be sent over RPC more efficiently.
+        args (tuple): the argument tuple for the ``func`` invocation.
+        kwargs (dict): is a dictionary of keyword arguments for the ``func``
+                       invocation.
+
+    Returns:
+        Returns the result of running ``func``on ``args`` and ``kwargs``.
+
+    Example::
+        On worker 0:
+        >>> import torch.distributed as dist
+        >>> dist.init_process_group(backend='gloo', rank=0, world_size=2)
+        >>> dist.init_model_parallel("worker0")
+        >>> ret = dist.rpc_sync("worker1", torch.add, args=(torch.ones(2), 3))
+        >>> dist.join_rpc()
+
+        On worker 1:
+        >>> import torch.distributed as dist
+        >>> dist.init_process_group(backend='gloo', rank=1, world_size=2)
+        >>> dist.init_model_parallel("worker1")
+        >>> dist.join_rpc()
+    """
+    fut = _invoke_rpc(to, func, args, kwargs)
+    return fut.wait()
+
+
+@_require_initialized
+def rpc_async(to, func, args=None, kwargs=None):
+    r"""
+    Make a non-blocking RPC call to run function ``func`` on worker ``to``. RPC
+    messages are sent and received in parallel to execution of Python code. This
+    method is thread-safe. This method will immediately return a
+    torch.distributed.FutureMessage that can be awaited on.
+
+    Arguments:
+        to (int or str): id or name of the destination worker.
+        func (callable): any callable function. builtin functions (like
+                         ``torch.add``) can be sent over RPC more efficiently.
+        args (tuple): the argument tuple for the ``func`` invocation.
+        kwargs (dict): is a dictionary of keyword arguments for the ``func``
+                       invocation.
+
+    Returns:
+        Returns a ``torch.distributed.FutureMessage`` object that can be waited
+        on. When completed, the return value of ``func`` on ``args`` and
+        ``kwargs`` can be retrieved from the ``FutureMessage`` object.
+
+    Example::
+
+        On worker 0:
+        >>> import torch.distributed as dist
+        >>> dist.init_process_group(backend='gloo', rank=0, world_size=2)
+        >>> dist.init_model_parallel("worker0")
+        >>> worker1 = dist.get_worker_id("worker1")
+        >>> fut1 = dist.rpc_async(worker1, torch.add, args=(torch.ones(2), 3))
+        >>> fut2 = dist.rpc_async(worker1, min, args=(1, 2))
+        >>> result = fut1.wait() + fut2.wait()
+        >>> dist.join_rpc()
+
+        On worker 1:
+        >>> import torch.distributed as dist
+        >>> dist.init_process_group(backend='gloo', rank=1, world_size=2)
+        >>> dist.init_model_parallel("worker1")
+        >>> dist.join_rpc()
+    """
+    fut = _invoke_rpc(to, func, args, kwargs)
+    return fut
+
+
 @_require_initialized
 def rpc(to, func, args=None, kwargs=None, async_call=False):
```

Comment:
Hey @rohan-varma, thanks for adding this. IIUC, the original goal was actually removing `rpc` and splitting it into `rpc_sync` and `rpc_async`. The reason for this is that, if `rpc` does not have a fixed return type, we won't be able to TorchScript it.

Reply:
@mrshenli Are we concerned about backwards compatibility here? I'm not completely familiar with the backwards compatibility protocols we follow, but won't completely removing `rpc` immediately break the code of whoever is using it when they update to the new version?

Reply:
Currently `test/test_rpc.py` and `test/test_dist_autograd.py` are the only dependency we have I believe, it should be sufficient to modify those accordingly. But I think your solution is more appropriate by marking `rpc` as deprecated. Let's land this PR and then remove `dist.rpc` later. :)